0xPolygon/polygon-edge
# Permission smart contract deployment

# Description

This PR implements permission smart contract deployment feature.

**How it works:**
The idea behind this feature is that we want to enable nodes to restrict who can deploy smart contracts. We've decided to add a deployment whitelist and only addresses inside this list can deploy smart contracts. When the deployment whitelist is empty anyone can deploy contracts 

To make this easy to use, we've added a few CLI commands:

1. AddAddress
`go run main.go whitelist deployment --add 0x5383Cb489FaCa92365Bb6f9f1FB40bD032E6365d --add 0x5383Cb489FaCa92365Bb6f9f1FB40bD032E6365d`
<img width="1680" alt="Screenshot 2022-08-22 at 19 01 35" src="https://user-images.githubusercontent.com/42786413/185977974-e2eee1ac-d155-4bae-bdab-9a653da6e370.png">

2. removeAddress
`go run main.go whitelist deployment --remove 0x5383Cb489FaCa92365Bb6f9f1FB40bD032E6365d --remove 0x5383Cb489FaCa92365Bb6f9f1FB40bD032E6365d`
<img width="1680" alt="Screenshot 2022-08-22 at 19 03 10" src="https://user-images.githubusercontent.com/42786413/185978190-fa9ffa2d-537e-4b8e-8edf-310f382bea04.png">

**After changing the deployment whitelist, nodes should be restarted to fetch the latest whitelist from genesis config**

3. Show all whitelist
`go run main.go whitelist show `
<img width="669" alt="Screenshot 2022-08-22 at 19 05 02 1" src="https://user-images.githubusercontent.com/42786413/185978454-bf12da82-2f80-4c0a-8c40-6a672b57619e.png">

# Changes include


# Checklist


## Testing


### Manual tests
Scenarios covered by unit tests:
1. Deployment whitelist is empty anyone can deploy:
    1. Start a network
    2. Deploy smart contract from any address

2. Addresses inside the whitelist can deploy smart contract
    1. Start a network
    2. Add the address to the whitelist
    3. Only that address should be able to deploy a smart contract

Other scenarios
1. User should be able to burn tokens on 0x address
2. Whitelist should not affect EOA transfers and smart contract interaction

# Documentation update

After the first round of reviews, I will write docs

# Additional comments

Fixes EDGE-735
